File: coauthor-for-vs-code/src/agent/modelHandlerOpenRouter.ts
// Standard library imports
// (none needed)

// Third-party imports
import OpenAI from 'openai';

// Local imports - log
import * as logger from '../logger/logUtils';

// Local imports - agent components
import { ModelHandlerOpenAI } from './modelHandlerOpenAI';
import { ToolState } from './ToolState';

const CHANNEL = 'Agent';
logger.initialize(CHANNEL);

/**
 * Handler for models accessed through OpenRouter.
 */
export class ModelHandlerOpenRouter extends ModelHandlerOpenAI {
  /** Returns OpenAI client configured with OpenRouter settings. */
  getClient(): OpenAI {
    const apiKey = this.getApiKey();
    const baseUrl = this.getBaseUrl();
    logger.info(CHANNEL, `Using OpenRouter API key. Base URL: ${baseUrl}`);
    return new OpenAI({
      apiKey,
      baseURL: baseUrl,
    });
  }

  /** Creates a response using OpenRouter's API with model-specific configuration. */
  async createResponse(
    client: OpenAI,
    messages: any[],
    temperature: number,
    systemPrompt?: string,
    endTag?: string,
  ): Promise<any> {
    const kwargs: any = {
      model: this.config.openrouterFullName, // Use OpenRouter model name
      messages,
      max_tokens: this.config.maxOutputTokens,
      temperature,
      extra_headers: { 'X-Title': 'CoA' },
    };

    if (endTag) {
      kwargs.stop = [endTag];
    }

    return client.chat.completions.create(kwargs);
  }
}

/**
 * Handler for Anthropic models using OpenAI-compatible API via OpenRouter.
 */
export class ModelHandlerAnthropicViaOpenRouter extends ModelHandlerOpenRouter {
  /** Returns OpenAI client configured with Anthropic's base URL. */
  getClient(): OpenAI {
    return new OpenAI({
      apiKey: this.getApiKey(),
      baseURL: this.getBaseUrl(),
    });
  }

  /** Updates message content with support for Anthropic's assistant prefill. */
  updateMessageContent(
    messages: any[],
    bestConnector: string,
    newResponse: string,
    toolState: ToolState,
  ): void {
    const lastMessage = messages[messages.length - 1];
    if (this.capabilities.supportsAssistantPrefill) {
      // although OpenAI models do not support assistant prefill, some models (such as Anthropic) via OpenRouter might do
      if (lastMessage.role === 'assistant') {
        if (Array.isArray(lastMessage.content)) {
          lastMessage.content[lastMessage.content.length - 1].text =
            bestConnector + newResponse;
        } else if (typeof lastMessage.content === 'string') {
          lastMessage.content = toolState.accumulatedOutput;
        }
      } else if (lastMessage.role === 'user') {
        messages.push({
          role: 'assistant',
          content: toolState.accumulatedOutput,
        });
      }
    }
  }
}

File: coauthor-python/coauthor/agent/model_handler_openrouter.py
"""Handlers for models accessed through OpenRouter."""

from typing import Any
from openai import OpenAI

from .model_handler_openai import OpenAIHandler
from .tool_state import ToolState

from ..logger import logger


class OpenRouterHandler(OpenAIHandler):
    """Handler for models accessed through OpenRouter."""

    def getClient(self) -> OpenAI:
        """Get OpenAI client with OpenRouter configuration."""
        api_key = self.get_api_key()
        base_url = self.get_base_url()
        logger.info(f"Using OpenRouter API key. Base URL: {base_url}")
        return OpenAI(
            api_key=api_key,
            base_url=base_url,
        )

    def createResponse(
        self,
        client: OpenAI,
        messages: list[dict],
        temperature: float,
        systemPrompt: str | None = None,
        endTag: str | None = None,
    ) -> Any:
        """Create a response using OpenRouter's API."""
        kwargs = {
            "model": self.config.openrouterFullName,  # Use OpenRouter model name
            "messages": messages,
            "max_tokens": self.config.maxOutputTokens,
            "temperature": temperature,
            "extra_headers": {"X-Title": "CoA"},
        }

        if endTag:
            kwargs["stop"] = [endTag]

        return client.chat.completions.create(**kwargs)


class AnthropicviaOpenrouterHandler(OpenRouterHandler):
    """Handler for Anthropic models using OpenAI-compatible API via OpenRouter."""

    def getClient(self) -> OpenAI:
        """Get OpenAI client with Anthropic's base URL."""
        return OpenAI(
            api_key=self.get_api_key(),
            base_url=self.get_base_url(),
        )

    def updateMessageContent(self, messages: list[dict], bestConnector: str, newResponse: str, toolState: ToolState) -> None:
        last_message = messages[-1]
        if self.capabilities.supportsAssistantPrefill:
            # although OpenAI models do not support assistant prefill, some models (such as Anthropic) via OpenRouter might do
            if last_message["role"] == "assistant":
                if isinstance(messages[-1]["content"], list):
                    messages[-1]["content"][-1]["text"] = bestConnector + newResponse
                elif isinstance(messages[-1]["content"], str):
                    messages[-1]["content"] = toolState.accumulatedOutput
            elif last_message["role"] == "user":
                messages.append({"role": "assistant", "content": toolState.accumulatedOutput})
